Lib/functools.py

```diff
@@ -192,6 +192,191 @@ def total_ordering(cls):
+        n2i = self.node2info
+        for node in result:
+            n2i[node].npredecessors = _NODE_OUT
+
+        # Clean the list of nodes that are ready and update
+        # the counter of nodes that we have returned.
+        self.ready_nodes.clear()
+        self.npassedout += len(result)
+
+        return result
+
+    def is_active(self):
+        """Return True if more progress can be made and ``False`` otherwise.
+
+        Progress can be made if cycles do not block the resolution and either there are still nodes ready
+        that haven't yet been returned by "get_ready" or the number of nodes marked "done" is less than the
+        number that have been returned by "get_ready".
+
+        Raises ValueError if called without calling "prepare" previously.
+        """
+        if self.ready_nodes is None:
+            raise ValueError("prepare() must be called first")
+        return self.nfinished < self.npassedout or bool(self.ready_nodes)
+
+
+    def done(self, node):
+        """Marks a nodes returned by "get_ready" as processed.
+
+        This method unblocks any successor of *node* for being returned in the future by a call to "get_ready"
+
+        Raises :exec:`ValueError` if *node* has already been marked as processed by a previous call to this
+        method or if *node* was not added to the graph by using "add" or if called without calling "prepare"
+        previously.
+        """
+
+        if self.ready_nodes is None:
+            raise ValueError("prepare() must be called first")
+
+        n2i = self.node2info
+
+        # Check if we know about this node (it was added previously using add()
+        if (nodeinfo := n2i.get(node)) is None:
+            raise ValueError(f"node {node!r} was not added using add()")
+
+        # If the node has not being returned (marked as ready) previously, inform the user.
+        stat = nodeinfo.npredecessors
+        if stat != _NODE_OUT:
+            if stat >= 0:
+                raise ValueError(f"node {node!r} was not passed out (still not ready)")
+            elif stat == _NODE_DONE:
+                raise ValueError(f"node {node!r} was already marked done")
+            else:
+                raise ValueError(f"node {node!r}: unknown status {stat}")
+
+        # Mark the node as processed
+        nodeinfo.npredecessors = _NODE_DONE
+
+        # Go to all the successors and reduce the number of predecessors, collecting all the ones
+        # that are ready to be returned in the next get_ready() call.
+        for successor in nodeinfo.successors:
+            successor_info = n2i[successor]
+            successor_info.npredecessors -= 1
+            if successor_info.npredecessors == 0:
+                self.ready_nodes.append(successor)
+        self.nfinished += 1
+
+    def _find_cycle(self):
+        todo = set(node for node in self.node2info)
+        for info in self.node2info.values():
+            todo |= set(info.successors)
+
```

Comment:
I'm still trying to convince myself this function is correct, and that the `if node in stack:` later can't lead to quadratic-time behavior (e.g., what about a long chain like 1 -> 2 -> 3 -> 4 -> 5 -> ...).  But, if it is correct, the three lines above could be replaced by

```python
    todo = set(self.node2info)
```

The way `node2info` is constructed guarantees that all nodes in successor lists are keys in `node2info`.

Reply:
> and that the `if node in stack:` later can't lead to quadratic-time behavior

If this is a concern we can introduce an extra `processed` set for the check

Reply:
> The way node2info is constructed guarantees that all nodes in successor lists are keys in node2info.

Added in 62a891d together with the set to protect the stack check.